packages/mongodb-dialects/java-driver/src/main/kotlin/com/mongodb/jbplugin/dialects/javadriver/glossary/NamespaceExtractor.kt

```diff
@@ -0,0 +1,410 @@
+                            )
+
+                        runCatching {
+                            Pair(
+                                assignment.concept,
+                                callToSuperConstructor!!.argumentList.expressions[indexOfParameter],
+                            )
+                        }.getOrNull()
+                    } else {
+                        Pair(assignment.concept, assignment.resolutionExpression)
+                    }
+                }
+            }
+
+        val collection = resolvedScopes.find { it.first == AssignmentConcept.COLLECTION }
+        val database = resolvedScopes.find { it.first == AssignmentConcept.DATABASE }
+        val client = resolvedScopes.find { it.first == AssignmentConcept.CLIENT }
+
+        if (collection != null && database == null) {
+            // if we have a parameter for the collection, but we don't have the database
+            // assume it's a call to getDatabase().getCollection()
+            return extractNamespaceFromDriverConfigurationMethodChain(collection.second as PsiMethodCallExpression)
+        } else if (collection != null && database != null) {
+            // if we have a parameter for a collection and database, try to resolve them either
+            // from the parent constructor or the actual constructor
+            val databaseString = database.second.tryToResolveAsConstantString()!!
+            val collectionString = collection.second.tryToResolveAsConstantString()!!
+            return Namespace(databaseString, collectionString)
+        } else if (client != null || resolvedScopes.size == 1) {
+            // if it's not a client and there is only one resolved variable
+            // guess from the actual constructor
+            val mongodbNamespaceDriverExpression =
+                currentClass.constructors.firstNotNullOfOrNull {
+                    val callToSuperConstructor =
+                        findChildrenOfType(it, PsiMethodCallExpression::class.java).first {
+                            it.methodExpression.text == "super" &&
+                                    it.methodExpression.resolve() == constructorAssignments.first().constructor
+                        }
+
+                    val indexOfParameter =
+                        constructorAssignments.first().constructor.parameterList.getParameterIndex(
+                            constructorAssignments.first().parameter!!,
+                        )
+                    callToSuperConstructor.argumentList.expressions[indexOfParameter]
+                }
+
+            return extractNamespaceFromDriverConfigurationMethodChain(
+                mongodbNamespaceDriverExpression as PsiMethodCallExpression,
+            )
+        }
+
+        return null
+    }
+
+    private fun resolveConstructorArgumentReferencesForField(
+        containingClass: PsiClass,
+        field: Pair<AssignmentConcept?, PsiField>,
+    ): List<FieldAndConstructorAssignment> {
```

Comment:
Just a general suggestion, but classes like these might be a bit easier to read if their methods provided example input/output pairs in doc comments

Reply:
Added a few comments on the more complex methods. I think you are right, the code is complex enough that we will forget the details at some point. Thanks for pointing it out!